### DISCUSSION TITLE
How can we reflect the optional/required json items in documentation

### DISCUSSION THREAD

**Discussion by silverfoxio** [fastapi/fastapi#8181] [Questions] [answered]
**Description**

How can I mark the optional/required json elements in the swagger documentation generated using FastAPI?
Currently, for POST verb, it just marked "* required" for the whole body and not individual json elements.

**Comment by haizaar** [comment]
Are you asking about optional/required fields of the JSON body?

**Comment by silverfoxio** [comment]
@haizaar - yes, that's about reflecting the optional/required json elements in the body for swagger documentation

**Comment by euri10** [comment, marked as answer]
It's working out of the box, if you use pydantic objects the generated schema will "translate" the requirements, for instance this pydantic object from the very beginning of the documentation (https://pydantic-docs.helpmanual.io/):
```
class UserM(BaseModel):
    id: int
    name = 'John Doe'
    signup_ts: datetime = None
    friends: List[int] = []
```
used this way:
```
@app.post("/userm")
async def userm(userm: UserM):
    pass
```
shows in the documentation as 

![Imgur](https://i.imgur.com/1vFf5Bq.png)

So quoting pydantic doc, you can check `id` is required (red star), `name` has a default so not-required (no red star) etc....
> What’s going on here:
id is of type int; the annotation only declaration tells pydantic that this field is required. Strings, bytes or floats will be coerced to ints if possible, otherwise an exception would be raised.
name is inferred as a string from the default, it is not required as it has a default.
signup_ts is a datetime field which is not required (None if it’s not supplied), pydantic will process either a unix timestamp int (e.g. 1496498400) or a string representing the date & time.
friends uses python’s typing system, it is required to be a list of integers, as with id integer-like objects will be converted to integers.

**Comment by euri10** [comment]
which makes me think reading the spec on 
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#requestBodyObject
that the generated schema defaults to True for the request body when the spec defaults to False, I guess a minor thinkg

**Comment by silverfoxio** [comment]
thanks, @euri10  and @haizaar

**Comment by haizaar** [comment]
> which makes me think reading the spec on
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#requestBodyObject
that the generated schema defaults to True for the request body when the spec defaults to False, I guess a minor thinkg

@euri10, What if you define the handler as `async def userm(userm: UserM = None):` (or any other default value)?

**Comment by euri10** [comment]
if you do that @haizaar the red "required" next to the Request Body disappears

**Comment by haizaar** [comment]
Which makes sense according to spec I assume, right?

On Tue, 16 Jul 2019 at 23:47, euri10 <notifications@github.com> wrote:

> if you do that @haizaar <https://github.com/haizaar> the red "required"
> next to the Request Body disappears
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tiangolo/fastapi/issues/388?email_source=notifications&email_token=AAAOGWKWBDHRVFEZ6MWYL2LP7XGNDA5CNFSM4IDVCTH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2A4ZBI#issuecomment-511822981>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAOGWJ3T34J5LV7OVBU5PTP7XGNDANCNFSM4IDVCTHQ>
> .
>


-- 
Zaar

**Comment by euri10** [comment]
yep indeed, seems like I got things reversed, like right and left, one day it'll be better :)

**Comment by tiangolo** [comment]
Hehe, thanks for the help @euri10 and @haizaar.

And thanks @sl2016 for reporting back and closing the issue.
